File: 프로그래머스/lv3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

vector<int> solution(vector<string> gems) {
    vector<int> answer(2);
    unordered_map<string, int> kind;
    unordered_map<string, int> dis;
    int num=0, cnt=0, dis_answer=gems.size()+1;
    string front=gems[0], end=gems[0];
    
    for(int i=0; i<gems.size(); i++){
        if(!kind[gems[i]]){
            kind[gems[i]]=i+1;
            num++;
        }
    }
    
    for(int i=0; i<gems.size(); i++){
        if(!dis[gems[i]]) cnt++;
        dis[gems[i]]=i+1;
        if(front==gems[i]){
            int s=dis.begin()->second;
            for(auto j=dis.begin(); j!=dis.end(); j++){
                if(j->second<=s){
                    front=j->first;
                    s=j->second;
                }
            }
        }
        end=gems[i];
        if(cnt==num){
            if(dis_answer>(dis[end]-dis[front])){
                answer[0]=dis[front];
                answer[1]=dis[end];
                dis_answer=dis[end]-dis[front];
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/lv3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;
class Solution {
    public int[] solution(String[] gems) {
        int[] answer = new int[2];
        Map<String, Integer> kind = new HashMap<>();
        for(int i=0; i<gems.length; i++){
            kind.put(gems[i], 0);
        }
        int s = 0, e = 0, n = kind.size(), as= -1, ae = -1;
        int cnt = 0;
        int len = Integer.MAX_VALUE;
        int t = 1;
        while(s<=e){
            if(cnt==n){
                if(len>e-s){
                    len = e-s;
                    as = s;
                    ae = e-1;
                }
                String st = gems[s];
                kind.replace(st, kind.get(st)-1);
                if(kind.get(st)==0) cnt--;
                s++;
            }
            else{
                if(e==gems.length) break;
                String gem = gems[e];
                if(kind.get(gem)==0){
                    kind.replace(gem, 1);
                    cnt++;
                } else{
                    kind.replace(gem, kind.get(gem)+1);
                }
                e++;
            }
        }
        
        answer[0] = as+1;
        answer[1] = ae+1;
        return answer;
    }
}